### PR TITLE
#74: Implement Post Filtering API

### LINKED ISSUE #74: [BE] API route for post filtering
## Context

We would like to make implement filtering on the server-side that gets called whenever a user makes changes to the post filters.

## Requirements

Create an API route that takes in an object of the selected filters. Use this object to filter the posts by their field values. Reference the `SelectedFilters` type in `Feed.tsx` if you would like to see what the frontend data looks like. You can also open the devtools console and change the filters on the site - we currently print the selected filters to console whenever a change is made. If you want, you can implement the backend so that the type is parsed down to its enum values (stripping the `label` key, array of enum values instead of array of objects, etc). Keep in mind that the posts should maintain a most to least recent ordering.

There are currently some unknowns that we (the LT) are considering - contact Katsuki for more info.

- e.g. if a post doesn’t have any temperament data, should all filters selected for it include that post? what is the default behavior for the filtering? etc

## Acceptance Criteria

- [ ]  Posts are filtered correctly based on the input criteria

### PR DESCRIPTION
**Pull request by tmthecoder**
## Implement Post Filtering API

Issue Number(s): #74 .

Adds an endpoint `getFilteredPosts` to provide a list of posts with the given filter

### Checklist

- [x] Requirements have been implemented
- [x] Acceptance criteria are met
- [x] Database schema docs have been updated or are not necessary
- [x] Code follows design and style guidelines
- [x] Commits follow guidelines (concise, squashed, etc)
- [x] Relevant reviewers (Senior Dev/EM/Designers) have been assigned to this PR

### Critical Changes

None

### Related PRs

None

### Testing

Query the API endpoint: `api/trpc/post.getFilteredPosts` with a URL-encoded query string from this JSON format: 

```json
{
    "breed": [breed-restrictions in array],
    "type": [type-restrictions in array],
    "age": [age-restrictions in array],
    "size": [size-restrictions in array],
    "gender": [gender-restrictions in array],
    "behavioral": [behavioral-restrictions in array],
    "goodWith": [Must be good-with (none if all is fine)],
    "houseTrained": single status, optional,
    "crateTrained": single status, optional
    "spayNeuterStatus": single status, optional,
    "spayNeuterStatus": single status
}
```

### CODE REVIEW

File: web/server/routers/post.ts [hunk 1]
```diff
@@ -61,6 +62,19 @@ const postSchema = z.object({
...
+const postFilterSchema = z.object({
+  type: z.array(z.nativeEnum(FosterType)),
+  breed: z.array(z.nativeEnum(Breed)),
+  age: z.array(z.nativeEnum(Age)),
+  size: z.array(z.nativeEnum(Size)),
+  gender: z.array(z.nativeEnum(Gender)),
+  goodWith: z.array(z.nativeEnum(GoodWith)),
+  behavioral: z.array(z.nativeEnum(Behavioral)),
+  houseTrained: z.nativeEnum(Trained).optional(),
+  crateTrained: z.nativeEnum(Trained).optional(),
```
Comment (katsukixyz) [inline web/server/routers/post.ts]: Had a conversation with the LT and we'll be removing crate trained altogether

**Review by katsukixyz — COMMENTED** [review]
Everything looks good! Crate trained can be kept in the post schema, but it shouldn't be a filterable option. Great work 👍

File: web/server/routers/post.ts (hunk 1 above)
Comment (tmthecoder) [reply web/server/routers/post.ts]: Should I remove it from both then and push that up?

Comment (katsukixyz) [reply web/server/routers/post.ts]: Remove it from the filter API route and the filter schema!

Comment (tmthecoder) [reply web/server/routers/post.ts]: Should be good now

**Review by katsukixyz — APPROVED** [review]
Thanks for the changes!

**Pull request merged**